data-avro/src/main/java/com/linkedin/data/avro/DataTranslator.java

```diff
@@ -609,18 +609,17 @@ private Object translate(Object value, DataSchema dataSchema, Schema avroSchema)
             Schema fieldAvroSchema = avroField.schema();
             Object fieldValue = map.get(fieldName);
             boolean isOptional = field.getOptional();
-            if (isOptional  || (
-                _dataTranslationOptions != null &&
-                ((DataMapToAvroRecordTranslationOptions) _dataTranslationOptions). getDefaultFieldDataTranslationMode()
-                    == PegasusToAvroDefaultFieldTranslationMode.DO_NOT_TRANSLATE))
+            if (isOptional)
             {
+              // process optional fields
```

Comment:
Generally LGTM, though the documentation here could be drastically improved. This would help to avoid long conversations in the future to figure out what's going on here. It looks like you already added a few, but here are my suggested edits:

> // process optional fields

This comment is unnecessary, since the `isOptional` condition above is clear.

Add a comment like this:

```
if (fieldDataSchema.getDereferencedType() != DataSchema.Type.UNION)
{
  // Translating from a non-union pegasus type to an embedded union type in Avro
  if (fieldValue == null)
```

> // optional field will be represented as union in Avro

This doesn't really explain what's happening here. I'd suggest:
```
// Extract the Avro type corresponding to the pegasus type from the Avro union
```

> // already a union

Should be something like:
```
// Translating from a pegasus union to an Avro union
```

> // if no fieldValue present, either NULL value or defaultValue will be assigned

Should be a little more descriptive:

```
// Required field is missing, assign default value if present
```

Then add another comment in the subcondition saying:

```
// Translate default value as null, depending on specified options
```